nexus/internal/filetransfer/file_transfer_manager.go

```diff
@@ -0,0 +1,223 @@
+	return func(fm *FileTransferManager) {
+		fm.logger = logger
+	}
+}
+
+func WithSettings(settings *service.Settings) FileTransferManagerOption {
+	return func(fm *FileTransferManager) {
+		fm.settings = settings
+	}
+}
+
+func WithFileTransfer(fileTransfer FileTransfer) FileTransferManagerOption {
+	return func(fm *FileTransferManager) {
+		fm.fileTransfer = fileTransfer
+	}
+}
+
+func (fm *FileTransferManager) fileCount() {
+	for fileType := range fm.fileCountsChan {
+		switch fileType {
+		case WandbFile:
+			fm.fileCounts.WandbCount++
+		case MediaFile:
+			fm.fileCounts.MediaCount++
+		case ArtifactFile:
+			fm.fileCounts.ArtifactCount++
+		default:
+			fm.fileCounts.OtherCount++
+		}
+	}
+	fm.wgfc.Done()
+}
+
+func NewUploadManager(opts ...FileTransferManagerOption) *FileTransferManager {
+
+	fm := FileTransferManager{
+		inChan:         make(chan interface{}, bufferSize),
+		fileCounts:     &service.FileCounts{},
+		fileCountsChan: make(chan FileType, bufferSize),
+		wgfc:           &sync.WaitGroup{},
+		wg:             &sync.WaitGroup{},
+		semaphore:      make(chan struct{}, defaultConcurrencyLimit),
+	}
+
+	for _, opt := range opts {
+		opt(&fm)
+	}
+
+	fm.wgfc.Add(1)
+	go fm.fileCount()
+
+	return &fm
+}
+
+// Start is the main loop for the fileTransfer
+func (fm *FileTransferManager) Start() {
+	fm.wg.Add(1)
+	go func() {
+		for task := range fm.inChan {
+			// add a task to the wait group
+			fm.wg.Add(1)
+			fm.logger.Debug("fileTransfer: got task", "task", task)
+			// spin up a goroutine per task
+			go func(task interface{}) {
+				// Acquire the semaphore
+				fm.semaphore <- struct{}{}
+				switch t := task.(type) {
+				case *UploadTask:
+					t.Err = fm.transfer(t)
+					// Release the semaphore
+					<-fm.semaphore
+					if t.Err != nil {
+						fm.logger.CaptureError(
+							"filetransfer: uploader: error uploading",
+							t.Err,
+							"path", t.Path, "url", t.Url,
+						)
+					}
+					// Execute the callback.
+					if t.CompletionCallback != nil {
+						t.CompletionCallback(t)
+					}
+
+					if t.Err == nil {
+						fm.fileCountsChan <- t.FileType
+					}
+				case *DownloadTask:
+					t.Err = fm.transfer(t)
+					// Release the semaphore
+					<-fm.semaphore
+					if t.Err != nil {
+						fm.logger.CaptureError(
+							"filetransfer: downloader: error downloading",
+							t.Err,
+							"path", t.Path, "url", t.Url,
+						)
+					}
+					// Execute the callback.
+					if t.CompletionCallback != nil {
+						t.CompletionCallback(t)
+					}
+
+					if t.Err == nil {
+						fm.fileCountsChan <- t.FileType
+					}
+				default:
+					fm.logger.Debug("fileTransfer: go routine for transfer task: invalid task type", "type", t)
+				}
+				// mark the task as done
+				fm.wg.Done()
+			}(task)
+		}
+		fm.wg.Done()
+	}()
+}
+
+// AddTask adds a task to the fileTransfer
+func (fm *FileTransferManager) AddTask(task interface{}) {
+	switch t := task.(type) {
```

Comment:
The two branches here are identical, why do we need the switch?

Reply:
Ah, I see. Thanks - I guess I was just trying to make sure the type of the task isn't invalid, but that'll get handled by a downstream check if needed. I'll update this.